shotover-proxy/src/message/mod.rs

```diff
@@ -20,325 +20,265 @@ use itertools::Itertools;
             Frame::None => Some(vec![]),
         }
     }
 
     pub fn invalidate_cache(&mut self) {
-        // TODO: invalidate cache when we have a cache to invalidate
-    }
-
-    #[must_use]
-    pub fn to_filtered_reply(&self) -> Self {
-        Message {
-            details: MessageDetails::Unknown,
-            modified: true,
-            original: match &self.original {
-                Frame::Redis(_) => Frame::Redis(RedisFrame::Error(
-                    "ERR Message was filtered out by shotover".into(),
-                )),
-                Frame::Cassandra(frame) => Frame::Cassandra(CassandraFrame {
-                    version: frame.version,
-                    stream_id: frame.stream_id,
-                    operation: CassandraOperation::Error(ErrorBody {
-                        error_code: 0,
-                        message: "Message was filtered out by shotover".into(),
-                        additional_info: AdditionalErrorInfo::Server,
-                    }),
-                    tracing_id: None,
-                    warnings: vec![],
+        // TODO: clear message details cache fields if we ever add any
+
+        self.inner = self.inner.take().map(|x| x.invalidate_cache());
+    }
+
+    // TODO: this could be optimized to avoid parsing the cassandra sql
+    pub fn to_filtered_reply(&mut self) -> Message {
+        Message::from_frame(match self.frame().unwrap() {
+            Frame::Redis(_) => Frame::Redis(RedisFrame::Error(
+                "ERR Message was filtered out by shotover".into(),
+            )),
+            Frame::Cassandra(frame) => Frame::Cassandra(CassandraFrame {
+                version: frame.version,
+                stream_id: frame.stream_id,
+                operation: CassandraOperation::Error(ErrorBody {
+                    error_code: 0x0,
+                    message: "Message was filtered out by shotover".into(),
+                    additional_info: AdditionalErrorInfo::Server,
                 }),
-                Frame::None => Frame::None,
-            },
-            return_to_sender: false,
-            meta_timestamp: None,
-        }
+                tracing_id: frame.tracing_id,
+                warnings: vec![],
+            }),
+            Frame::None => Frame::None,
+        })
     }
 
-    /// Gets the `QueryType` for a message.
-    /// First checks the `details` of the message and then falls back to the `original` message.
-    pub fn get_query_type(&self) -> QueryType {
-        if let MessageDetails::Query(query) = &self.details {
-            query.query_type.clone()
-        } else {
-            self.original.get_query_type()
+    pub fn get_query_type(&mut self) -> QueryType {
+        match self.frame() {
+            Some(Frame::Cassandra(cassandra)) => cassandra.get_query_type(),
+            Some(Frame::Redis(redis)) => redis_query_type(redis), // free-standing function as we cant define methods on RedisFrame
```

Comment:
Do you think it's worth writing an abstraction similar to `CassandraFrame` for `RedisFrame`? Then we could have a trait that has all the methods that we need for a frame to be used in message.

Though at this point I think we should just get this merged. Would be good to consider for later if we end up needing more frame types added.

Reply:
Assuming you mean write a wrapper around redis-protocol types and call it RedisFrame, I cant see that being necessary.
Cassandra has a lot more going on and needed more complicated frame for shotover transforms to work with nicely.
Redis is a lot simpler and so the RedisFrame provided by redis-protocol should be fine.